Sprint 14

superaydumov opened the pull request:
Pull request according to Yandex  Praktikum course (sprint 14).

superaydumov committed:
Created TrackersViewController and StatisticViewController programmatically (added UIDatePicker, UISearchTextField, UINavigationBar and UITabBarController).
Added needed assets from Figma template.
Refactored UIColor+Extension to avoid problems with colors representation.

superaydumov committed:
Added cancelSearchTextFieldButton and extension for TrackersViewController (to conform UITextFieldDelegate).
Refactored constraintsSetup method and added additionalConstraintsSetup method.

superaydumov committed:
Added TrackerCreatorViewController to navigate tracker creation (habit or non-regular event).
Replaced files for better navigation in project.
Refactored TrackersViewController and StatisticsViewController.

superaydumov committed:
Add EventCreatorViewController:
- create buttons, labels, text field;
- logic to show different titles, labels and constraints depending on event type is added;
- UITextField restriction for 38 symbols is added.

Little tweaks in code (TrackersViewController and TrackerCreatorViewController methods and properties).

superaydumov committed:
Refactored UIDatePicker in TrackersViewController class (created computed property for datePicker and moved all it's settings inside).

superaydumov committed:
CategoryViewController and ScheduleViewController are added.
TrackersViewController and StatisticViewController were refactored (changed constraint value and add setWeekDay method to TrackersViewController).
Add logic to EventCreatorViewController to show CategoryViewController and ScheduleViewController.
Add WeekDay enum to maintain adding days in ScheduleViewController.
Add ScheduleTableViewCellDelegate protocol to append days in schedule array when UISwitch is on.

superaydumov committed:
Changed background color in Assets.
ScheduleViewControllerDelegate is added to pass schedule from ScheduleViewController to EventCreatorViewController.
Added delegate property to cell (ScheduleTableViewCell) in ScheduleViewController.

superaydumov committed:
Tracker, TrackerCategory and TrackerRecord structures are added to the project.

superaydumov committed:
Created TrackersCollectionViewCell and implemented it in TrackersViewController.

superaydumov committed:
Created TrackersCellSuplementaryView and implemented it in TrackerViewController.

superaydumov committed:
Created configureCell method in TrackersCollectionViewCell class.
Configured cell in TrackersViewController and implemented configureCell method.
Created TrackersCollectionViewCellDelegate to maintain checkButton tap (implemented protocol in TrackersViewController and created completedTrackers method).

superaydumov committed:
Created MockData class and implemented it in TrackersViewController's categories property to show trackers in sections.

Created EventCreatorViewControllerDelegate and TrackerCreatorViewControllerDelegate with createTracker method to maintain adding new tracker. Implemented these protocols in EventCreatorViewController and TrackerCreatorViewController. Also added createTracker method in TrackersViewController.

Refactored code in TrackersCellSuplementaryView and TrackersCollectionViewCell.

superaydumov committed:
Added yeadMonthDay components in extension to Date() to maintain comparing date from datePicker and current date.

superaydumov committed:
Added more trackers in MockData class to check scroll and maintaining sections in collectionView.
Added numberOfDay property to WeekDay enumeration to use Int in updateCategories method.
Created updateCategories and numbersOfSections methods in TrackersViewController.

superaydumov committed:
Little changes in code.

amardanov commented on Tracker/MockData.swift:
```diff
@@ -0,0 +1,57 @@
+
+final class MockData {
+    static var categories: [TrackerCategory] = [
+        TrackerCategory(categoryName: "Развлечения",
```
Рекомендую придерживаться следующего стиля написания длинных вызовов и длинных объявлений функций. Как пример можешь использовать то, что я напишу. Такой формат принят за основу во многих компаниях и следование ему даст несколько очков к карме при работе над боевыми проектами

```
TrackerCategory(
            categoryName: "Развлечения",
            trackers: [
                Trackers(
                    id: UUID(),
                    name: "Играть в прятки как ниндзя",
                    color: .colorSelection2,
                    emoji: "🥷🏻",
                    schedule: [.monday, .wednesday, .friday]
                ),
                Trackers(
                    id: UUID(),
                    name: "Стать самураем",
                    color: .colorSelection12,
                    emoji: "🐥",
                    schedule: [.tuesday, .thursday, .saturday]
                )
            ]
        )
```

amardanov commented on Tracker/Trackers/EventCreatorViewController.swift:
```diff
@@ -0,0 +1,355 @@
+    @objc func createButtonDidTap(sender: AnyObject) {
+        let tracker = Trackers(id: UUID(), name: textField.text ?? "Без названия", color: .colorSelection7, emoji: "🥌", schedule: schedule)
+        delegate?.createTracker(tracker: tracker, categoryName: "Радостные мелочи")
+        print("tracker created in EventCreatorViewController")
```
Принты лучше не оставлять в коде приложения. Так код будет чище и ревьюеру будет проще проверять за счет уменьшения количества строк

amardanov commented on Tracker/Trackers/EventCreatorViewController.swift:
```diff
@@ -0,0 +1,355 @@
+
+    // MARK: - UITextField extension
+
+extension UITextField {
```
Лучше выносить такие расширения в отдельный файл. В этом файле можно оставить только расширения самого контроллера. Как выше, например

amardanov commented on Tracker/Trackers/TrackersViewController.swift:
```diff
@@ -0,0 +1,375 @@
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        let count = visibleCategories.count
+        if count == 0 {
+            collectionView.isHidden = true
```
collectionView.isHidden = visibleCategories.count == .zero

superaydumov committed:
Refactored MockData and TrackersViewController class after reviewer's comments.
Moved UITextField extension to Helpers folder.
Created and implemented createButtonUpdate method in EventCreatorViewController class to maintain createButton behaviour.

superaydumov committed:
Created scheduleSubtitle in EventCreatorViewController to show days which were selected in ScheduleViewController.
Updated createSchedule method to add days in schedule string.

superaydumov committed:
Updated createButtonDidTap method in EventCreatorViewController to maintain adding regular and non-regular trackers
